iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -559,167 +582,45 @@ class TransposeGenericDotGeneral : public OpRewritePattern<mhlo::DotGeneralOp> {
-    };
-
-    int lhsContractingDimIndex, rhsContractingDimIndex, lhsParallelDimIndex,
-        rhsParallelDimIndex;
-    SmallVector<int64_t, 4> lhsNewShape, rhsNewShape;
-    std::tie(lhsNewShape, lhsContractingDimIndex, lhsParallelDimIndex) =
-        computeCollapsedShape(lhsShapeType.getShape(), lhsBatchingDims,
-                              lhsContractingDims);
-
-    std::tie(rhsNewShape, rhsContractingDimIndex, rhsParallelDimIndex) =
-        computeCollapsedShape(rhsShapeType.getShape(), rhsBatchingDims,
-                              rhsContractingDims);
-    SmallVector<int64_t, 4> resultNewShape = {lhsNewShape[0],
-                                              lhsNewShape[lhsParallelDimIndex],
-                                              rhsNewShape[rhsParallelDimIndex]};
-    Type dotGeneralResultType =
-        RankedTensorType::get(resultNewShape, resultType.getElementType());
+    if (lhs == op.lhs() && rhs == op.rhs()) return failure();
 
-    auto loc = op.getLoc();
-    Value reshapedLhs = rewriter.create<mhlo::ReshapeOp>(
-        loc, RankedTensorType::get(lhsNewShape, lhsShapeType.getElementType()),
-        op.lhs());
-    Value reshapedRhs = rewriter.create<mhlo::ReshapeOp>(
-        loc, RankedTensorType::get(rhsNewShape, rhsShapeType.getElementType()),
-        op.rhs());
     auto dimensionNumbers = mhlo::DotDimensionNumbersAttr::get(
-        rewriter.getContext(),
-        /*lhs_batching_dimensions=*/{0},
-        /*rhs_batching_dimensions=*/{0},
-        /*lhs_contracting_dimensions=*/{lhsContractingDimIndex},
-        /*rhs_contracting_dimensions=*/
-        {rhsContractingDimIndex});
-    Value dotGeneralResult = rewriter.create<mhlo::DotGeneralOp>(
-        loc, dotGeneralResultType, reshapedLhs, reshapedRhs, dimensionNumbers,
+        rewriter.getContext(), /*lhs batching*/ 0, /*rhs batching*/ 0,
+        /*lhs contraction*/ 2, /*rhs contraction*/ 1);
+    auto lhsNewType = lhs.getType().cast<RankedTensorType>();
+    auto rhsNewType = rhs.getType().cast<RankedTensorType>();
+    bool needReshapeResult = lhsNewType.getRank() < lhsShapeType.getRank() ||
+                             rhsNewType.getRank() < rhsShapeType.getRank();
+    // batching、lhs parallel、lhs parallel
+    SmallVector<int64_t, 4> newShape = {lhsNewType.getShape()[0],
+                                        lhsNewType.getShape()[1],
+                                        rhsNewType.getShape()[2]};
+    auto newResultType =
+        needReshapeResult
+            ? RankedTensorType::get(newShape, resultType.getElementType())
+            : op.getType();
+
+    Value result = rewriter.create<mhlo::DotGeneralOp>(
+        op.getLoc(), newResultType, lhs, rhs, dimensionNumbers,
         op.precision_configAttr());
 
-    Value result =
-        rewriter.create<mhlo::ReshapeOp>(loc, resultType, dotGeneralResult);
+    if (needReshapeResult) {
+      result = rewriter.create<mhlo::ReshapeOp>(
```

Comment:
Okay, I think it's fine. Because [doc](https://www.tensorflow.org/xla/operation_semantics#dotgeneral) and your link say that.

> It follows that the resulting dimension number starts with the batch dimension, then the 'lhs' non-contracting/non-batch dimension, and finally the 'rhs' non-contracting/non-batch dimension.